awsenv/replacer.go

```diff
@@ -1,89 +1,180 @@
 package awsenv
 
 import (
+	"context"
+	"errors"
 	"os"
-	"strings"
 
 	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/request"
 	"github.com/aws/aws-sdk-go/service/ssm"
-	"github.com/pkg/errors"
+	"golang.org/x/sync/errgroup"
 )
 
-// ParameterGetter is implemented by the ssm client
+const batchLimit = 10
+
+var (
+	setenv  = os.Setenv
+	environ = os.Environ
+)
+
+// ParameterGetter is implemented by the ssm client.
 type ParameterGetter interface {
 	GetParameter(*ssm.GetParameterInput) (*ssm.GetParameterOutput, error)
 }
 
-// Replacer handles replacing existing environment variables with parameter values
-type Replacer struct {
-	ssm    ParameterGetter
-	prefix string
+// ParametersGetter is more efficient than ParameterGetter and is also
+// implemented by the ssm client.
+type ParametersGetter interface {
+	GetParametersWithContext(aws.Context, *ssm.GetParametersInput, ...request.Option) (*ssm.GetParametersOutput, error)
 }
 
 // NewReplacer returns a Replacer that will operate on env vars with the
-// given value prefix, using the given ParameterGetter.
+// given value prefix, using the given ParamsGetter. If ssm also implements
+// ParametersGetter, that interface will be used instead.
+//
 func NewReplacer(envValuePrefix string, ssm ParameterGetter) *Replacer {
```

Comment:
Whats the reasoning for not just taking a `ParametersGetter` here instead of wrapping it? Since this repo wasn't originally intended to be a library, I think its okay to make some breaking changes here (and this change shouldn't break the CLI contract).

Reply:
Indeed-ish. If we're talking about breaking the lib API, we should look at the [support-sdk-v2](https://github.com/sendgrid/aws-env/blob/support-sdk-v2/awsenv/replacer.go) branch, since trading one hyper-specific method for another still leaves us with something brittle, hence the decision to not break anything in this PR (or break it completely, but in a non-brittle way in that other branch).